sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala

```diff
@@ -71,6 +71,9 @@ private[spark] object SQLConf {
   // Whether to perform partition discovery when loading external data sources.  Default to true.
   val PARTITION_DISCOVERY_ENABLED = "spark.sql.sources.partitionDiscovery.enabled"
 
+  // Whether to perform partition column type inference. Default to false.
+  val PARTITION_COLUMN_TYPE_INFERENCE = "spark.sql.sources.partitionColumnTypeInference.enabled"
```

Comment:
We probably want to set this to `true` to keep the default behavior backwards compatible with Spark 1.3.x.